python/GafferSceneUI/ShaderQueryUI.py

```diff
@@ -0,0 +1,426 @@
+
+	if plug is None :
+		return suffix
+
+	prefix = ""
+
+	if parentPlug is not None :
+		n = plug.node()
+		idx = n.getChildIndex( n[parentPlug], plug )
+
+		prefix = n["queries"][idx]["name"].getValue()
+
+		currentPlug = plug
+		while (
+			currentPlug is not None and
+			currentPlug != n[parentPlug][idx]
+		) :
+
+			suffix = "." + currentPlug.getName() + suffix
+			currentPlug = currentPlug.parent()
+
+	result = prefix + suffix
+	result = result.replace( ".", "_" ) if sanitise else result
+
+	return result
+
+
+##########################################################################
+# Metadata
+##########################################################################
+
+Gaffer.Metadata.registerNode(
+
+	GafferScene.ShaderQuery,
+
+	"description",
+	"""
+	Queries shader parameters from a scene location, creating outputs
+	for each parameter.
+	""",
+
+	plugs = {
+
+		"scene" : [
+
+			"description",
+			"""
+			The scene to query the shader for.
+			""",
+
+		],
+
+		"location" : [
+
+			"description",
+			"""
+			The location within the scene to query the shader at.
+			> Note : If the location does not exist then the query will not be
+			> performed and all outputs will be set to their default values.
+			""",
+
+			"plugValueWidget:type", "GafferSceneUI.ScenePathPlugValueWidget",
+			"scenePathPlugValueWidget:scene", "scene",
+			"nodule:type", "",
+
+		],
+
+		"shader" : [
+
+			"description",
+			"""
+			The name of the shader to query.
+			> Note : If the shader does not exist then the query will not be
+			> performed and all outputs will be set to their default values.
+			""",
+
+			"plugValueWidget:type", "GafferUI.PresetsPlugValueWidget",
+			"presetsPlugValueWidget:allowCustom", True,
+
+			"preset:None", "",
+			"nodule:type", "",
+
+		],
+
+		"inherit" : [
+
+			"description",
+			"""
+			Should inherited shader attributes or be considered or not.
+			""",
+
+			"nodule:type", "",
+
+		],
+
+		"queries" : [
+
+			"description",
+			"""
+			The shader parameters to be queried - arbitrary numbers of shader
+			parameters may be added as children of this plug via the user interface,
+			or via python. Each child is a `NameValuePlug` whose `name` plug is
+			the shader parameter to query, and whose `value` plug is the default
+			value to use if the shader parameter can not be retrieved.
+
+			The full network of the shader given by `shader` is available
+			to be queried. Parameters on shaders in the network other than the
+			output shader can be specified as `shaderName.parameterName`.
+
+			> Note : If either the shader or parameter does not exist then the
+			> query will not be performed and all outputs will be set to their
+			> default values.
+			""",
+
+			"plugValueWidget:type", "GafferUI.LayoutPlugValueWidget",
+
+			"layout:customWidget:footer:widgetType", "GafferSceneUI.ShaderQueryUI._ShaderQueryFooter",
+			"layout:customWidget:footer:index", -1,
+
+			"nodule:type", "GafferUI::CompoundNodule",
```

Comment:
The nodule for the names should be hidden but I was thinking having a nodule for the default would make it more intuitive to connect an expression.

But now I'm thinking it might make it too attractive for someone to connect a shader output to, so perhaps it should be hidden.

Reply:
Removed the nodules in 6da2cd62dd76b0ba050248530a7f41e9173f3e7a